pkg/util/debouncer/debouncer.go

```diff
@@ -0,0 +1,311 @@
+	if opts.KeyFunc == nil {
+		return nil, errors.New("keyFunc is required")
+	}
+
+	if opts.ProcessHandler == nil {
+		return nil, errors.New("processHandler is required")
+	}
+
+	if opts.ErrorHandler == nil {
+		opts.ErrorHandler = func(_ T, _ error) {}
+	}
+
+	if opts.Reg == nil {
+		opts.Reg = prometheus.NewRegistry()
+	}
+
+	return &Group[T]{
+		opts:           opts,
+		buffer:         make(chan T, opts.BufferSize),
+		debouncers:     make(map[string]*debouncer[T]),
+		metrics:        newMetrics(opts.Reg, opts.Name),
+		processHandler: opts.ProcessHandler,
+		errorHandler:   opts.ErrorHandler,
+	}, nil
+}
+
+func (g *Group[T]) Add(value T) error {
+	select {
+	case g.buffer <- value:
+		g.metrics.itemsAddedCounter.Inc()
+		return nil
+	default:
+		g.metrics.itemsDroppedCounter.Inc()
+		return ErrBufferFull
+	}
+}
+
+func (g *Group[T]) Start(ctx context.Context) {
+	g.ctx, g.cancel = context.WithCancel(ctx)
+	g.wg.Add(1)
+	go func() {
+		defer g.wg.Done()
+		for {
+			select {
+			case <-g.ctx.Done():
+				return
+			case value := <-g.buffer:
+				g.processValue(value)
+			}
+		}
+	}()
+}
+
+func (g *Group[T]) Stop() {
+	if g.cancel != nil {
+		g.cancel()
+		g.wg.Wait()
+	}
+}
+
+func (g *Group[T]) processValue(key T) {
+	keyStr := g.opts.KeyFunc(key)
+	g.debouncersMu.Lock()
+	deb, ok := g.debouncers[keyStr]
+	if !ok {
+		deb = newDebouncer[T](g.opts.MinWait, g.opts.MaxWait, key, func(v T) {
+			g.processWithMetrics(g.ctx, v, g.processHandler)
+
+			g.debouncersMu.Lock()
+			defer g.debouncersMu.Unlock()
+			if current, exists := g.debouncers[keyStr]; exists && current == deb {
+				delete(g.debouncers, keyStr)
+			}
+		})
+		g.wg.Add(1)
+		go func() {
+			defer g.wg.Done()
+			deb.run(g.ctx)
+		}()
+		g.debouncers[keyStr] = deb
+	}
+	g.debouncersMu.Unlock()
+
+	deb.reset()
+}
+
+func (g *Group[T]) processWithMetrics(ctx context.Context, value T, processFunc ProcessFunc[T]) {
+	timer := prometheus.NewTimer(g.metrics.processingDurationHistogram)
+	defer timer.ObserveDuration()
+	g.metrics.itemsProcessedCounter.Inc()
+
+	if err := processFunc(ctx, value); err != nil {
+		g.errorHandler(value, err)
+		g.metrics.processingErrorsCounter.Inc()
+	}
+}
+
+// debouncer handles debouncing for a specific key.
+type debouncer[T any] struct {
+	key         T
+	resetChan   chan struct{}
+	minWait     time.Duration
+	maxWait     time.Duration
+	processFunc func(T)
+}
+
+// newDebouncer creates a new key debouncer.
+func newDebouncer[T any](minWait, maxWait time.Duration, key T, processFunc func(T)) *debouncer[T] {
+	deb := &debouncer[T]{
+		key:         key,
+		resetChan:   make(chan struct{}, 1),
+		minWait:     minWait,
+		maxWait:     maxWait,
+		processFunc: processFunc,
+	}
+	return deb
+}
+
+// reset triggers a timer reset for the minWait.
+func (d *debouncer[T]) reset() {
+	select {
+	case d.resetChan <- struct{}{}:
+		// Value sent successfully.
+	default:
+		// Value was dropped. Is not an issue as
+		// a reset is already about to being processed.
```

Comment:
It can also be already being processed, in which case it's unclear whether more calls to `Add` are supposed to "debounce" or trigger new processing. I would suggest to clarify it in the doc.

Reply:
Added some more docs, also on `Add`.